crates/ethereum/evm/src/execute.rs

```diff
@@ -66,84 +69,76 @@ where
+    /// Reference to the parent factory providing access to [`SystemCaller`] and [`ChainSpec`].
+    #[deref]
+    #[deref_mut]
+    factory: &'a EthExecutionStrategyFactory<EvmConfig>,
+
+    /// Block being executed.
+    block: &'a SealedBlock,
+    /// The EVM used by strategy.
+    evm: EvmWithHook<Evm>,
+    /// Receipts of executed transactions.
+    receipts: Vec<Receipt>,
 }
 
-impl<DB, EvmConfig> EthExecutionStrategy<DB, EvmConfig>
-where
-    EvmConfig: Clone,
-{
+impl<'a, Evm, EvmConfig> EthExecutionStrategy<'a, Evm, EvmConfig> {
     /// Creates a new [`EthExecutionStrategy`]
-    pub fn new(state: State<DB>, chain_spec: Arc<ChainSpec>, evm_config: EvmConfig) -> Self {
-        let system_caller = SystemCaller::new(evm_config.clone(), chain_spec.clone());
-        Self { state, chain_spec, evm_config, system_caller }
+    pub fn new(
+        evm: Evm,
+        block: &'a SealedBlock,
+        factory: &'a EthExecutionStrategyFactory<EvmConfig>,
+    ) -> Self {
+        Self { evm: EvmWithHook::new(evm), factory, block, receipts: Vec::new() }
     }
 }
 
-impl<DB, EvmConfig> BlockExecutionStrategy for EthExecutionStrategy<DB, EvmConfig>
+impl<'db, DB, E, EvmConfig> BlockExecutionStrategy for EthExecutionStrategy<'_, E, EvmConfig>
 where
-    DB: Database,
-    EvmConfig: ConfigureEvm<
-        Header = alloy_consensus::Header,
-        Transaction = reth_primitives::TransactionSigned,
-    >,
+    DB: Database + 'db,
+    E: Evm<DB = &'db mut State<DB>, Tx = EvmConfig::TxEnv>,
+    EvmConfig: ConfigureEvmEnv<Transaction = TransactionSigned>,
 {
-    type DB = DB;
     type Error = BlockExecutionError;
     type Primitives = EthPrimitives;
 
-    fn apply_pre_execution_changes(
-        &mut self,
-        block: &RecoveredBlock<reth_primitives::Block>,
-    ) -> Result<(), Self::Error> {
+    fn apply_pre_execution_changes(&mut self) -> Result<(), Self::Error> {
         // Set state clear flag if the block is after the Spurious Dragon hardfork.
-        let state_clear_flag = self.chain_spec.is_spurious_dragon_active_at_block(block.number());
-        self.state.set_state_clear_flag(state_clear_flag);
-
-        let mut evm = self.evm_config.evm_for_block(&mut self.state, block.header());
-
-        self.system_caller.apply_pre_execution_changes(block.header(), &mut evm)?;
+        let state_clear_flag =
+            self.chain_spec.is_spurious_dragon_active_at_block(self.block.number());
+        self.evm.db_mut().set_state_clear_flag(state_clear_flag);
+        self.system_caller.apply_pre_execution_changes(self.block.header(), &mut self.evm)?;
 
         Ok(())
     }
 
     fn execute_transactions(
         &mut self,
-        block: &RecoveredBlock<reth_primitives::Block>,
-    ) -> Result<ExecuteOutput<Receipt>, Self::Error> {
-        let mut evm = self.evm_config.evm_for_block(&mut self.state, block.header());
-
+        transactions: &[TransactionSigned],
+        senders: &[Address],
```

Comment:
should this take an iter iter over `Recovered<&Tx>` instead?

we're copying the sender in this function anyway

https://github.com/paradigmxyz/reth/blob/d64fece1ca8bf03f079480465119b6a0ddcb59f7/crates/primitives-traits/src/block/recovered.rs#L294-L300